### PR TITLE
[DBZ-1254] Re-order the captured tables

### PR DESCRIPTION
**Pull request by willome**
Re-order the captured tables in the same order as the 'table.whitelist' parameter.

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java [hunk 1]
```diff
@@ -216,6 +220,26 @@ private void delaySnapshotIfNeeded(ChangeEventSourceContext context) throws Inte
...
+        return tableIds
+                .stream()
+                .filter(tid -> id.asPredicate().test(tid.toString()))
+                .findFirst();
+    }
+
+    private Set<TableId> sort(Set<TableId> capturedTables) throws Exception {
+        String value = connectorConfig.getConfig().getString(RelationalDatabaseConnectorConfig.TABLE_WHITELIST);
+        if (value != null) {
+            return Strings.setOfRegex(value, Pattern.CASE_INSENSITIVE)
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: Could you please take a look at Strings.listOfRegex() method? IMHO you want to keep the order of regexes

File: debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java [hunk 2]
```diff
@@ -216,6 +220,26 @@ private void delaySnapshotIfNeeded(ChangeEventSourceContext context) throws Inte
      */
     protected void connectionCreated(SnapshotContext snapshotContext) throws Exception {
     }
+    
+    private Optional<TableId> toTableId(Set<TableId> tableIds, Pattern id) {
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: What if more than one table matches? IMHO you yhoul return stream of TableIds and then `flatMap` it

File: debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java [hunk 3]
```diff
@@ -232,7 +256,7 @@ private void determineCapturedTables(SnapshotContext ctx) throws Exception {
             }
         }
 
-        ctx.capturedTables = capturedTables;
+        ctx.capturedTables = sort(capturedTables);
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: What if one pattern matches more than one table? Would the order be deterministic? Maybe the captured tables should be also sorted to ensure exact order for such case.

Comment (willome) [reply debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: I am sorry I do not fully understand what you mean. I had an Integration test with some overlapping patterns. Do you think it is enough and answer your question ?

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: If you will have a single regex that will capute tables aa, ab, ac - could you predict in which order they'll be captured?

Comment (willome) [reply debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: Actually no. Do you think we should sort by String "natural order" ? For me it is not really relevant but maybe there are some use cases that I do not see right now.

Comment (willome) [reply debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: I could just sort the stream like that :
````
private Stream<TableId> toTableIds(Set<TableId> tableIds, Pattern pattern) {
        return tableIds
                .stream()
                .filter(tid -> pattern.asPredicate().test(tid.toString()))
                .sorted();
    }
````

Comment (willome) [reply debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: PR updated.

Comment (willome) [reply debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java]: PR updated : capturesTables is sorted by natural order (even if table.whitelist is missing).

**Comment by jpechane** [comment]
@willome Applied, thanks!

**Pull request merged**